processor/labelsprocessor/labels_processor.go

```diff
@@ -0,0 +1,124 @@
+			return fmt.Errorf("Labels Processor configuration contains an empty key or value")
+		}
+	}
+
+	//Ensure no duplicate keys exist
+	keys := make(map[string]bool)
+	for _, elem := range cfg.Labels {
+		_, value := keys[elem.Key]
+		if value {
+			return fmt.Errorf("Labels Processor configuration contains duplicate keys")
+		}
+		keys[elem.Key] = true
+	}
+
+	return nil
+}
+
+func (lp *labelMetricProcessor) ProcessMetrics(_ context.Context, md pdata.Metrics) (pdata.Metrics, error) {
+
+	otlpMetrics := pdata.MetricsToOtlp(md)
+
+	for _, otlpMetric := range otlpMetrics {
+		for _, instrMetric := range otlpMetric.GetInstrumentationLibraryMetrics() {
+			for _, metric := range instrMetric.GetMetrics() {
+
+				// Multiple types of Data Points exists, and each of them must be handled differently
+				if metric.GetIntSum() != nil {
+					intDataPoints := metric.GetIntSum().GetDataPoints()
+					handleIntDataPoints(intDataPoints, lp)
+				} else if metric.GetIntGauge() != nil {
+					intDataPoints := metric.GetIntGauge().GetDataPoints()
+					handleIntDataPoints(intDataPoints, lp)
+				} else if metric.GetDoubleGauge() != nil {
+					doubleDataPoints := metric.GetDoubleGauge().GetDataPoints()
+					handleDoubleDataPoints(doubleDataPoints, lp)
+				} else if metric.GetDoubleSum() != nil {
+					doubleDataPoints := metric.GetDoubleSum().GetDataPoints()
+					handleDoubleDataPoints(doubleDataPoints, lp)
+				} else if metric.GetIntHistogram() != nil {
+					intHistogramDataPoints := metric.GetIntHistogram().GetDataPoints()
+					handleIntHistogramDataPoints(intHistogramDataPoints, lp)
+				} else if metric.GetDoubleHistogram() != nil {
+					doubleHistogramDataPoints := metric.GetDoubleHistogram().GetDataPoints()
+					handleDoubleHistogramDataPoints(doubleHistogramDataPoints, lp)
+				}
+
+			}
+		}
+	}
+
+	return md, nil
+}
+
+func handleIntDataPoints(intDataPoints []*v1.IntDataPoint, lp *labelMetricProcessor) {
+	for _, label := range lp.cfg.Labels {
+		for _, dataPoint := range intDataPoints {
+			deDuplicateAndAppend(&dataPoint.Labels, label.Key, label.Value)
+		}
+	}
+}
+
+func handleDoubleDataPoints(doubleDataPoints []*v1.DoubleDataPoint, lp *labelMetricProcessor) {
+	for _, label := range lp.cfg.Labels {
+		for _, dataPoint := range doubleDataPoints {
+			deDuplicateAndAppend(&dataPoint.Labels, label.Key, label.Value)
+		}
+	}
+}
+
+func handleIntHistogramDataPoints(intHistogramDataPoints []*v1.IntHistogramDataPoint, lp *labelMetricProcessor) {
+	for _, label := range lp.cfg.Labels {
+		for _, dataPoint := range intHistogramDataPoints {
+			deDuplicateAndAppend(&dataPoint.Labels, label.Key, label.Value)
+		}
+	}
+}
+
+func handleDoubleHistogramDataPoints(doubleHistogramDataPoints []*v1.DoubleHistogramDataPoint, lp *labelMetricProcessor) {
+	for _, label := range lp.cfg.Labels {
+		for _, dataPoint := range doubleHistogramDataPoints {
+			deDuplicateAndAppend(&dataPoint.Labels, label.Key, label.Value)
+		}
+	}
+}
```

Comment:
I wonder if we can create an interface with `GetLabels` method to get away with declaring 4 methods that are doing essentially the same thing. 

I recall that Go is flexible in the sense that you can declare an interface, and as long as a type has the method with same signature, you can use the interface to reference the type.

Reply:
I was quite annoyed myself because I couldn't think of a way off the top of my head at the time to reduce the amount of code here. I will look into creating an interface.

Reply:
After some playing around I could not find a way to leverage interfaces and avoid repetition of code

Reply:
I also looked into how the Prometheus Remote Write Exporter author's handled this, and it seems like they also have similar logic with multiple functions doing the exact same thing, but with different data type inputs: https://github.com/open-telemetry/opentelemetry-collector/blob/master/exporter/prometheusremotewriteexporter/helper.go#L279

Reply:
ok, that's too bad then.